### ISSUE TITLE
gene-panel-data/fetch really slow on impact and 30 genes (oncoprint)

### ISSUE THREAD

**Issue by inodb** [cBioPortal/cbioportal#4015]
http://www.cbioportal.org/index.do?cancer_study_id=msk_impact_2017&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&case_set_id=msk_impact_2017_cnaseq&gene_list=AKR1C3%2520AR%2520CYB5A%2520CYP11A1%2520CYP11B1%2520CYP11B2%2520CYP17A1%2520CYP19A1%2520CYP21A2%2520HSD17B1%2520HSD17B10%2520HSD17B11%2520HSD17B12%2520HSD17B13%2520HSD17B14%2520HSD17B2%2520HSD17B3%2520HSD17B4%2520HSD17B6%2520HSD17B7%2520HSD17B8%2520HSD3B1%2520HSD3B2%2520HSD3B7%2520RDH5%2520SHBG%2520SRD5A1%2520SRD5A2%2520SRD5A3%2520STAR&geneset_list=+&tab_index=tab_visualize&Action=Submit&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna

The content download takes almost 10 seconds

Reported by: https://groups.google.com/forum/#!topic/cbioportal/tpsOYX8zlfg

**Comment by inodb** [comment]
<img width="1058" alt="screen shot 2018-03-21 at 5 12 07 pm" src="https://user-images.githubusercontent.com/1334004/37737709-05b9f3f6-2d2b-11e8-9572-894a5bb95487.png">

**Comment by inodb** [comment]
@adamabeshouse @jjgao @zhx828 Discussed with @ersinciftci . This seems to be a result of: https://github.com/cBioPortal/cbioportal/issues/3628.

The endpoint gives 10K * 30 = 300K objects, which seems overkill.

It prolly makes more sense to link the gene panel to a genomic profile no? That way you only need one panel per profile

**Comment by zhx828** [comment]
@inodb it's the similar situation in study view Ersin and I have been discussing. @ersinciftci in oncoprint I think it uses the unique ids?

**Comment by adamabeshouse** [comment]
@inodb im not sure how else to do it, there are 300K pieces of information.. we could reduce the size of the returned object but thats not much of a reduction esp considering gzip

**Comment by inodb** [comment]
@adamabeshouse So I looked a bit at the schema - one thing we can make use of is that the molecular profile / sample id is always unique. So one can remove the `* 30` factor above (for each gene).

https://github.com/cBioPortal/cbioportal/blob/master/db-scripts/src/main/resources/cgds.sql#L350

**Comment by inodb** [comment]
Not sure exactly what happens in content download in chrome profiler, but it takes almost 10s (seems weird for a file of just 1.3MB, see above), maybe the gzip overhead?

**Comment by ersinciftci** [comment]
@inodb I don't think this user's specific problem is about this API though, frontend rendering takes 45 seconds after all the APIs finish returning the response, and the page becomes unresponsive during that time, and sometimes crashes.

**Comment by adamabeshouse** [comment]
@inodb you still need to give one piece of information per gene per sample, no matter what, right? I'm pretty sure the way gzip works that redundancy (sample id, profile id) will be compressed anyway

**Comment by jjgao** [comment]
One option would be loading the data in two pieces:

1. gene panel id per sample
2. gene panel meta including the list of genes per gene panel

based on both data, you can determine whether or not a gene is sequenced for a sample.

**Comment by jjgao** [comment]
Another option would be returning the data in a different format, .e.g, in a matrix
```
{
"Genes": ["g1", "g2", ...],
"Samples":["s1", "s2", "s3", ...]
"sequenced": [
  [true, true, false, ...], // g1
  [false, true, true, ...], // g2
  ...
]
}
```

You could even try to code them in bits to reduce the response size and maybe also not returning Genes and Samples back...

**Comment by jjgao** [comment]
... and this is a really critical one. So let's focus on it tomorrow!

**Comment by inodb** [comment]
@ersinciftci right - that's a good point. The endpoint is only about 10 seconds of those 45 seconds. We should have another issue for dealing with oncoprint performance once all data is there.

@adamabeshouse I agree that gzip prolly takes care of the redundancy, but something is taking 10 seconds to download. It might be the gzip overhead either on backend or frontend, so if that's the case it's still a worthwhile exercise reducing the data we're sending.

In terms of how you represent the data it def matters in terms of memory print. The way it's done now is `nr_genes * nr_samples`. But really you only need `nr_samples + (nr_genes * nr_gene_panels)`.

I like the first option ☝️

**Comment by ersinciftci** [comment]
I downloaded the 109MB response and used `gzip file.txt` to compress (creates a 1.3MB file.txt.gz file) and `gzip -d file.txt.gz` in terminal, both operations happen instantly, so doesn't seem like gzip has a big overhead, but might be different for browsers and for streaming data.

Assuming gzip does what it promises, there should be no difference between @jj's second option, a matrix, and the current response, as they ultimately has the same amount of data, excluding the repeating strings (which gzip should take care of, hence 109MB becomes 1.3MB). And @jj's first option was what we were already doing before https://github.com/cBioPortal/cbioportal/issues/3628

So why downloading 1.3MB data takes 10 seconds is still mystery to me, similar to what we saw in study view, I did some googling but couldn't find much, but as I said before, frontend takes 3 times more time than APIs in this user's case, and sometimes crashes, so maybe first we should check if the frontend takes a lot of time because the gene panel data is big, or because there are some parts need optimization. So we can keep this issue open, and open another one for oncoprint frontend, and keep track of them separately.

**Comment by ersinciftci** [comment]
One option is we can deploy the 1.11 version (it doesn't include https://github.com/cBioPortal/cbioportal/issues/3628 but still has the new oncoprint) to staging and see if the same query takes much less time, if so, we can just revert https://github.com/cBioPortal/cbioportal/issues/3628

**Comment by inodb** [comment]
@ersinciftci i just tried v1.11.3 and it takes about 1.3s to get gene panel data with my local machine using query below. Compared to 11s in 1.12.0. So I think it makes sense to switch back if easy? (@adamabeshouse ). Frontend still takes quite a bit, not sure if there's a big difference there.

http://www.cbioportal.org/index.do?cancer_study_id=msk_impact_2017&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&case_set_id=msk_impact_2017_cnaseq&gene_list=AKR1C3%2520AR%2520CYB5A%2520CYP11A1%2520CYP11B1%2520CYP11B2%2520CYP17A1%2520CYP19A1%2520CYP21A2%2520HSD17B1%2520HSD17B10%2520HSD17B11%2520HSD17B12%2520HSD17B13%2520HSD17B14%2520HSD17B2%2520HSD17B3%2520HSD17B4%2520HSD17B6%2520HSD17B7%2520HSD17B8%2520HSD3B1%2520HSD3B2%2520HSD3B7%2520RDH5%2520SHBG%2520SRD5A1%2520SRD5A2%2520SRD5A3%2520STAR&geneset_list=+&tab_index=tab_visualize&Action=Submit&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna

Also - I don't think it makes sense to send over redundant data just because gzip will handle it. In this particular case we're sending over a matrix of 300K points (10K samples * 30 genes), even though you only need 10K samples + (30 genes * 2 gene panels).

But yeah we should figure out why sending over 1.3MB takes so long. Maybe it has more to do with mysql dealing with these large datasets while sending it over

**Comment by jjgao** [comment]
I'll propose to do the following. If all agree, let's create individual issues and discuss from there?

1. revert back to the previous way of retrieving gene panel data in the frontend, ie. loading gene panel meta data and sample data separately @adamabeshouse 
2. profile frontend performance of oncoprint once 1 is done @adamabeshouse (maybe @zhx828 can also help)
3. reimplement gene panel end points in the new api @ersinciftci (@adamabeshouse @zhx828 @inodb @jjgao can help redeisn the api)
4. (not very urgent) investigate whether gzipping has a big overhead

Anything I missed?

**Comment by ersinciftci** [comment]
https://github.com/cBioPortal/cbioportal/pull/4019 reverts https://github.com/cBioPortal/cbioportal/pull/3793 and https://github.com/cBioPortal/cbioportal/pull/3766, so we should have similar performance to 1.11.3 after this is merged and frontend is also reverted.

**Comment by ersinciftci** [comment]
Created another PR https://github.com/cBioPortal/cbioportal/pull/4020, this one doesn't revert, it changes to a similar structure but still keeps `wholeExomeSequenced`
